File: go/crypto/glue.go
// Package cryptobox provides the Go side of stablecoin-forge's cryptography:
// the deterministic "glue" (canonical JSON, SHA-384, HKDF-SHA3, AES-256-GCM) that
// must agree byte-for-byte with the Rust provider, plus ML-KEM key establishment
// and a FIPS-mode check. Everything here is Go standard library only, so the
// module builds with no third-party dependencies.
package cryptobox

import (
	"bytes"
	"crypto/aes"
	"crypto/cipher"
	"crypto/hkdf"
	"crypto/sha3"
	"crypto/sha512"
	"encoding/json"
	"fmt"
)

// CanonicalJSON encodes v as canonical JSON: compact separators, struct field
// order preserved, HTML escaping disabled (to match Python/Rust output). This is
// the exact byte string that gets signed.
func CanonicalJSON(v any) ([]byte, error) {
	var buf bytes.Buffer
	enc := json.NewEncoder(&buf)
	enc.SetEscapeHTML(false)
	if err := enc.Encode(v); err != nil {
		return nil, err
	}
	return bytes.TrimRight(buf.Bytes(), "\n"), nil
}

// Fingerprint returns the SHA-384 (FIPS 180-4) digest of data.
func Fingerprint(data []byte) []byte {
	sum := sha512.Sum384(data)
	return sum[:]
}

// HKDFSHA3_256 derives keyLength bytes via HKDF (RFC 5869) with SHA3-256
// (FIPS 202). An empty salt means the RFC's all-zero salt of hash length.
func HKDFSHA3_256(ikm, salt []byte, info string, keyLength int) ([]byte, error) {
	return hkdf.Key(sha3.New256, ikm, salt, info, keyLength)
}

// AES256GCMSeal encrypts plaintext with AES-256-GCM (NIST SP 800-38D). The nonce
// must be 12 bytes; the 16-byte tag is appended to the ciphertext.
func AES256GCMSeal(key, nonce, aad, plaintext []byte) ([]byte, error) {
	if len(key) != 32 {
		return nil, fmt.Errorf("AES-256 key must be 32 bytes, got %d", len(key))
	}
	if len(nonce) != 12 {
		return nil, fmt.Errorf("GCM nonce must be 12 bytes, got %d", len(nonce))
	}
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, err
	}
	return gcm.Seal(nil, nonce, plaintext, aad), nil
}

// AES256GCMOpen is the inverse of AES256GCMSeal.
func AES256GCMOpen(key, nonce, aad, ctTag []byte) ([]byte, error) {
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, err
	}
	return gcm.Open(nil, nonce, ctTag, aad)
}

File: crates/invar-crypto/src/glue.rs
//! Deterministic "glue" around standard primitives — the composition layer that
//! must agree byte-for-byte across languages. Each function maps to a recipe in
//! `conformance/vectors.json`.

use aes_gcm::aead::{Aead, KeyInit, Payload};
use aes_gcm::{Aes256Gcm, Nonce};
use hkdf::Hkdf;
use invar_core::error::{InvarError, Result};
use sha2::{Digest, Sha384};
use sha3::Sha3_256;

/// Canonical JSON encoding (RFC 8259): compact separators, key order preserved
/// (serde_json is built with `preserve_order`). This is the exact byte string
/// that gets signed.
///
/// Note: matches Python `json.dumps(separators=(",",":"))`. Non-ASCII is emitted
/// as UTF-8 (not `\uXXXX`); keep signing preimages ASCII for cross-stack parity.
pub fn canonical_json(value: &serde_json::Value) -> Result<Vec<u8>> {
    serde_json::to_vec(value).map_err(|e| InvarError::Serialization(e.to_string()))
}

/// SHA-384 content fingerprint (FIPS 180-4).
pub fn fingerprint(data: &[u8]) -> Vec<u8> {
    let mut h = Sha384::new();
    h.update(data);
    h.finalize().to_vec()
}

/// HKDF (RFC 5869) with SHA3-256 (FIPS 202) as the hash. `salt` empty means the
/// RFC's all-zero salt of hash-length.
pub fn hkdf_sha3_256(ikm: &[u8], salt: &[u8], info: &[u8], len: usize) -> Result<Vec<u8>> {
    let hk = Hkdf::<Sha3_256>::new(Some(salt), ikm);
    let mut okm = vec![0u8; len];
    hk.expand(info, &mut okm)
        .map_err(|e| InvarError::Crypto(format!("hkdf expand: {e}")))?;
    Ok(okm)
}

/// AES-256-GCM seal (NIST SP 800-38D). 12-byte nonce; the 16-byte tag is appended
/// to the ciphertext (the `aes-gcm` crate's default framing).
pub fn aes256gcm_seal(key: &[u8], nonce: &[u8], aad: &[u8], plaintext: &[u8]) -> Result<Vec<u8>> {
    if key.len() != 32 {
        return Err(InvarError::Crypto("AES-256 key must be 32 bytes".into()));
    }
    if nonce.len() != 12 {
        return Err(InvarError::Crypto("GCM nonce must be 12 bytes".into()));
    }
    let cipher = Aes256Gcm::new(key.into());
    cipher
        .encrypt(
            Nonce::from_slice(nonce),
            Payload {
                msg: plaintext,
                aad,
            },
        )
        .map_err(|e| InvarError::Crypto(format!("aes-gcm seal: {e}")))
}

/// AES-256-GCM open (inverse of [`aes256gcm_seal`]).
pub fn aes256gcm_open(key: &[u8], nonce: &[u8], aad: &[u8], ct_tag: &[u8]) -> Result<Vec<u8>> {
    let cipher = Aes256Gcm::new(key.into());
    cipher
        .decrypt(Nonce::from_slice(nonce), Payload { msg: ct_tag, aad })
        .map_err(|e| InvarError::Crypto(format!("aes-gcm open: {e}")))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn aead_bad_key_or_nonce_length_errors() {
        assert!(aes256gcm_seal(&[0u8; 16], &[0u8; 12], b"", b"x").is_err()); // key too short
        assert!(aes256gcm_seal(&[0u8; 32], &[0u8; 8], b"", b"x").is_err()); // nonce wrong
    }

    #[test]
    fn aead_open_rejects_tampered_ciphertext() {
        let key = [7u8; 32];
        let nonce = [3u8; 12];
        let mut sealed = aes256gcm_seal(&key, &nonce, b"", b"secret").unwrap();
        let last = sealed.len() - 1;
        sealed[last] ^= 0xff; // corrupt the tag
        assert!(aes256gcm_open(&key, &nonce, b"", &sealed).is_err());
    }
}
